csrc/fusion_segmenter.cpp

```diff
@@ -316,16 +426,152 @@ std::unique_ptr<SegmentedFusion> SegmentedFusion::fromCompleteFusion(
   single_group->setHeuristic(heuristic);
   single_group->setID(0);
 
+  // Used to log the number of values and expressions in the fusion for
+  // serialization sanity check.
+  segmented_fusion_ptr->finalize();
   return segmented_fusion_ptr;
 }
 
 SegmentedFusion::SegmentedFusion(std::unique_ptr<Fusion> fusion)
     : segmented_fusion_name_{segmentedFusionName()},
       impl_(this),
-      complete_fusion_(std::move(fusion)) {
+      complete_fusion_(std::move(fusion)),
+      initial_vals_size_{0},
+      initial_exprs_size_{0} {
   annotateFP16IntermediateTensors();
 }
 
+flatbuffers::Offset<serde::SegmentedFusion> SegmentedFusion::serialize(
+    flatbuffers::FlatBufferBuilder& builder) const {
+  FUSER_PERF_SCOPE("SegmentedFusion::serialize");
+  const std::unordered_map<Val*, int64_t>& vals_to_id_map =
+      completeFusion()->deterministic_vals_map();
+  const std::unordered_map<Expr*, int64_t>& exprs_to_id_map =
+      completeFusion()->deterministic_exprs_map();
+  const std::unordered_map<SegmentedGroup*, int64_t>& groups_map =
+      impl_.groups_map();
+  const std::unordered_map<SegmentedEdge*, int64_t>& edges_map =
+      impl_.edges_map();
+
+  std::vector<flatbuffers::Offset<serde::SegmentedEdge>> edges_fb;
+  edges_fb.reserve(edges_.size());
+  for (SegmentedEdge* se : edges_) {
+    edges_fb.push_back(serialize(builder, se, vals_to_id_map, groups_map));
+  }
+
+  std::vector<flatbuffers::Offset<serde::SegmentedGroup>> groups_fb;
+  groups_fb.reserve(groups_.size());
+  for (SegmentedGroup* sg : groups_) {
+    groups_fb.push_back(sg->serialize(
+        builder, vals_to_id_map, exprs_to_id_map, groups_map, edges_map));
+  }
+
+  std::vector<int64_t> force_fp16_tv_fb;
+  force_fp16_tv_fb.reserve(force_fp16_tv_set_.size());
+  for (auto tv : force_fp16_tv_set_) {
+    force_fp16_tv_fb.push_back(vals_to_id_map.at(tv));
+  }
+
+  return serde::CreateSegmentedFusionDirect(
+      builder,
+      segmented_fusion_name_,
+      initial_vals_size_,
+      initial_exprs_size_,
+      &edges_fb,
+      &groups_fb,
+      &force_fp16_tv_fb,
+      toUnderlying(std::get<PrimDataType>(force_half_precision_type_.type)));
+}
+
+void SegmentedFusion::deserialize(const serde::SegmentedFusion* buffer) {
+  FUSER_PERF_SCOPE("SegmentedFusion::deserialize");
+  NVF_ERROR(buffer != nullptr, "serde::SegmentedFusion is nullptr.");
```

Comment:
You can avoid this check by passing in a const ref.

Reply:
All FlatBuffer fields are optional and are returned as pointers, so I don't think we can avoid a check.